datafusion/sql/src/planner.rs

```diff
@@ -587,40 +589,70 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
             })
     }
 
-    pub(crate) fn convert_data_type(&self, sql_type: &SQLDataType) -> Result<DataType> {
+    pub(crate) fn convert_data_type_to_field(
+        &self,
+        sql_type: &SQLDataType,
+    ) -> Result<FieldRef> {
         // First check if any of the registered type_planner can handle this type
         if let Some(type_planner) = self.context_provider.get_type_planner() {
             if let Some(data_type) = type_planner.plan_type(sql_type)? {
-                return Ok(data_type);
+                return Ok(Field::new("", data_type, true).into());
             }
         }
 
         // If no type_planner can handle this type, use the default conversion
         match sql_type {
             SQLDataType::Array(ArrayElemTypeDef::AngleBracket(inner_sql_type)) => {
                 // Arrays may be multi-dimensional.
-                let inner_data_type = self.convert_data_type(inner_sql_type)?;
-                Ok(DataType::new_list(inner_data_type, true))
+                let inner_data_type = self.convert_data_type_to_field(inner_sql_type)?;
+
+                // Lists are allowed to have an arbitrarily named field;
+                // however, a name other than 'item' will cause it to fail an
+                // == check against a more idiomatically created list in
+                // arrow-rs which causes issues. We use the low-level
+                // constructor here to preserve extension metadata from the
+                // child type.
+                Ok(Field::new(
```

Comment:
It would be nice to start wrapping this pattern into helper methods / extension traits . 


I started playing around and was able to move a lot of the mechanics around and simpify them:
- https://github.com/paleolimbot/datafusion/pull/1